test/gtest/uct/test_pending.cc

```diff
@@ -577,3 +578,130 @@ UCS_TEST_SKIP_COND_P(test_uct_pending, pending_fairness,
+        modify_config("RC_FC_ENABLE", "n", true);
+    }
+
+    pending_send_request_t* pending_alloc_short() {
+        pending_send_request_t *req = new pending_send_request_t();
+        req->ep                     = m_e1->ep(0);
+        req->uct.func               = pending_send_short_op;
+        return req;
+    }
+
+    static ucs_status_t am_handler_check_rx_order(void *arg, void *data,
+                                                  size_t length, unsigned flags) {
+        volatile bool *comp_received = (volatile bool*)arg;
+        uint64_t hdr                 = *(uint64_t*)data;
+
+        /* We expect that message sent from pending callback will arrive
+         * before the one sent from the completion callback. */
+        if (hdr == m_pending_hdr) {
+            m_pend_received = true;
+            EXPECT_FALSE(*comp_received);
+        } else if (hdr == m_comp_hdr) {
+            *comp_received = true;
+            EXPECT_TRUE(m_pend_received);
+        } else {
+            EXPECT_EQ(m_am_hdr, hdr);
+        }
+
+        return UCS_OK;
+    }
+
+    static ucs_status_t pending_send_short_op(uct_pending_req_t *self) {
+        pending_send_request_t *req = ucs_container_of(self,
+                                                       pending_send_request_t,
+                                                       uct);
+
+        return uct_ep_am_short(req->ep, AM_ID, m_pending_hdr, NULL, 0);
+    }
+
+    static void completion_cb(uct_completion_t *self, ucs_status_t c_status) {
+        am_completion_t *comp = ucs_container_of(self, am_completion_t, uct);
+
+        EXPECT_UCS_OK(c_status);
+
+        ucs_status_t status = uct_ep_am_short(comp->ep, AM_ID, m_comp_hdr,
+                                              NULL, 0);
+        EXPECT_TRUE(!UCS_STATUS_IS_ERR(status) ||
+                    (status == UCS_ERR_NO_RESOURCE));
+    }
+
+protected:
+    static const uint64_t m_am_hdr;
+    static const uint64_t m_pending_hdr;
+    static const uint64_t m_comp_hdr;
+    static const uint8_t  AM_ID;
+    static bool m_pend_received;
+};
+
+bool test_uct_pending_ooo::m_pend_received         = false;
+const uint64_t test_uct_pending_ooo::m_am_hdr      = 0x0ul;
+const uint64_t test_uct_pending_ooo::m_pending_hdr = 0x1ul;
+const uint64_t test_uct_pending_ooo::m_comp_hdr    = 0x2ul;
+const uint8_t test_uct_pending_ooo::AM_ID          = 10;
+
+/* Check that pending requests are processed before the sends from
+ * completion callbacks */
+UCS_TEST_SKIP_COND_P(test_uct_pending_ooo, send_ooo_with_comp,
+                     !check_caps(UCT_IFACE_FLAG_AM_SHORT |
+                                 UCT_IFACE_FLAG_AM_ZCOPY |
+                                 UCT_IFACE_FLAG_PENDING))
+{
+    initialize();
+
+    bool comp_received = false;
+    m_pend_received    = false;
+
+    install_handler_sync_or_async(m_e2->iface(), AM_ID,
+                                  am_handler_check_rx_order, &comp_received);
```

Comment:
i don't think this test really supports CB_ASYNC because it checks for order, but if some am is received on progress thread and another on main thread the order may not be guaranteed